runtime/parachains/src/hrmp.rs

```diff
@@ -470,24 +531,35 @@ pub mod pallet {
 		///
 		/// If there are pending HRMP close channel requests, you can use this
 		/// function process all of those requests immediately.
-		#[pallet::weight(0)]
-		pub fn force_process_hrmp_close(origin: OriginFor<T>) -> DispatchResult {
+		///
+		/// Total number of closing channels must be provided as witness data.
+		#[pallet::weight(<T as Config>::WeightInfo::force_process_hrmp_close(*_channels))]
+		pub fn force_process_hrmp_close(origin: OriginFor<T>, _channels: u32) -> DispatchResult {
 			ensure_root(origin)?;
 			Self::process_hrmp_close_channel_requests();
 			Ok(())
 		}
 
-		/// This cancels a pending open channel request. It can be canceled be either of the sender
+		/// This cancels a pending open channel request. It can be canceled by either of the sender
 		/// or the recipient for that request. The origin must be either of those.
 		///
 		/// The cancellation happens immediately. It is not possible to cancel the request if it is
 		/// already accepted.
-		#[pallet::weight(0)]
+		///
+		/// Total number of open requests (i.e. `HrmpOpenChannelRequestsList`) must be provided as
+		/// witness data.
+		#[pallet::weight(<T as Config>::WeightInfo::hrmp_cancel_open_request(*open_requests))]
 		pub fn hrmp_cancel_open_request(
 			origin: OriginFor<T>,
 			channel_id: HrmpChannelId,
+			open_requests: u32,
```

Comment:
This actually may complicate the usage of the API. Those calls are issued by parachains via UMP and they are necessarily less flexible, I imagine. (E.g. issued by governance)

Reply:
Understandable, but that's just the way it is, if you want to have accurate (and not overly pessimistic) weights. We do the same in many other substrate APIs, and the transaction submission app (aka. a _UI_) will take care of it for you. For example, to do `withdraw_unbonded`, you have to pass in the number of slashing spans. That is also complicating the API, but it is intention. 

In this case, I believe the parachains can also deal with this extra parameter. If you really insist, the only other way is to put a `const` max upper bound for open request. This won't be trivial either, as the number of open request is not strictly bounded in the code.